### PR TITLE
Add capabiltity to specify sources for release image content

### PR DESCRIPTION
**Pull request by abhinavdahiya**
##### add new fields to allow specifying the sources for release image
```go
type InstallConfig struct {
    ///
    // ReleaseImageSources is the list of sources/repositories for pulling the release-image
    // content.
    // Any of the source/repository can be used to pull the release-image content.
    // The order is the list has no meaning in terms of preference.
    ReleaseImageSources []ReleaseImageSource `json:"releaseImageSources"`
    ///
}
// ReleaseImageSource defines a source for pulling the release-image content.
type ReleaseImageSource struct {
    Repository string `json:"repository"`
}
```
##### setup the bootstrap node with mirrors `/etc/containers/registries.conf`

* Adds a new asset releaseimage.Source that uses the ReleaseImageSources to calculate the sources for release-image.
* Adds the containers-registries.conf template to create the registry sections [1], when more than one sources have been configured for the release-image.
  For a list of sources, the containers-registries.conf is setup so that each source can use all the other sources in the list as the mirror.
  for example, if the list of sources is a, b, c, d, then the containers-registries.conf is setup like:
  registry = a, mirrors = b, c, d
  regsitry = b, mirrors = a, c, d
  registry = c, mirrors = a, b, d
  registry = d, mirrors = a, b, c

  For each registry only mirror-by-digest is set to true.

##### push the image content source policy object for release image

* Create a `ImageContentSourcePolicy` cluster scoped object with name `release-image`
* Uses the releaseimage.Sources asset to generate the `ImageContentSourcePolicy`'s  `RepositoryDigestMirrors` for the release-image.

### CODE REVIEW

File: docs/dev/alternative_release_image_sources.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+## Details
+
+### ImageContentSourcePolicy
+
+If a list of `ReleaseImageSources` is specified, the list of digest based [RepositoryDigestMirrors][repository-digest-mirrors] includes,
+
+* the embedded release-image `repository`, and
+* all the items in the `ReleaseImageSources`
+
+**IMPORTANT**: there's an assumption that all the containers image references in the release-image are on the same repository for the embedded release-image.
```
Comment (smarterclayton) [inline docs/dev/alternative_release_image_sources.md]: Note that today the release image and the release image content are not in the same repository (quay.io/openshift-release-dev/ocp-release and ocp-4.0-art-dev-latest).  In the future we anticipate moving the content to other locations transparently as we update (to different repos or registries), and we may want to change that after install time, rather than before (i.e. we may not want to hardcode that in the installer, since that is within the payload).

File: docs/dev/alternative_release_image_sources.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+## Controlling the content
+
+The content of the `release-image`, i.e. the digest, continues to the controlled by the embedded release-image location or the `OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE` env.
+
+## Controlling the source
+
+The installer allows the users to specify a list of `ReleaseImageSources` through the InstallConfig, which become the **authoritative** list of container image repositories that can be used to fetch the `release-image` content which includes the release-image and all the included image references for the various container images for operators.
+
+## Details
```
Comment (smarterclayton) [inline docs/dev/alternative_release_image_sources.md]: Maybe add a section in here "How to populate the ReleaseImageSources field", and my suggestion would be that we have the mirror command create the correct values that a user cuts and pastes.  That allows release mirror to calculate all the values.

File: docs/dev/alternative_release_image_sources.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+### release-image location
+
+The release-image location that is propagated to the bootstrap node and the cluster-version-operator will continue to be the embedded release-image location if **NO** `ReleaseImageSources` were specified.
+
+If a list of `ReleaseImageSources` is specified, the release-image location that is propagated to the bootstrap node and the cluster-version-operator will comprise of the repository from the list (preferentially the first one), and the content (i.e the digest) from the embedded release-image.
+
+### Bootstrap machine containers-registries.conf
+
+If a list of `ReleaseImageSources` is specified, the [Registries][registry-containers-registries-conf] will be configured to have each source be a mirror for another.
```
Comment (smarterclayton) [inline docs/dev/alternative_release_image_sources.md]: Should we consider calling this `ReleaseImageSourceRepositories` or is that too wordy?

File: docs/dev/alternative_release_image_sources.md [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+repositoryDigestMirrors:
+- sources:
+  - q.io/ocp/release-x.y
+  - local.registry.com/ocp/release-x.y
+- sources:
+  - q.io/openshift/x.y
+  - local.registry.com/ocp/release-x.y
+- sources:
+  - q.io/rhhi/x.y
+  - local.registry.com/ocp/release-x.y
```
Comment (umohnani8) [inline docs/dev/alternative_release_image_sources.md]: If this is an ordered list, don't we want the primary repository to be listed at the end so that the user knows that the mirrors will be tried first and primary last.
Also, since `repositoryDigestMirrors` is just a list of sources, we don't really know what the primary repo is, does it mean `registries.conf` should have an entry for each source listed as the primary? Something like if `repoistoryDigestMirrors: A, B, C, D` then registries.conf will have the following entries
```
[[registry]]location = A; mirrors = B, C, D
[[registry]]location = B; mirrors = A, C, D
[[registry]]location = C; mirrors = A, B, D
[[registry]]location = D; mirrors = A, B, C
```
cc @mtrmac @mrunalp

File: docs/dev/alternative_release_image_sources.md [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+## release-image content vs release-image source
+
+`release-image` content is what operators get installed and what version of each operator (i.e. the container image) gets installed to the cluster. While, `release-image` source is **where** the `release-image` content gets pulled.
+
+Currently the installer controls **both** the `release-image` content and the `release-image` source using the embedded release-image location or the `OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE` env, but to allow users to use a release-image from their private registry (change the `release-image` source), but keep the `release-image` content identical to one vetted by OpenShift, the installer needs to allow specifying source separate from the content.
+
+## Controlling the content
+
+The content of the `release-image`, i.e. the digest, continues to the controlled by the embedded release-image location or the `OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE` env.
```
Comment (umohnani8) [inline docs/dev/alternative_release_image_sources.md]: should be "continues to be controlled"

File: docs/dev/alternative_release_image_sources.md (hunk 4 above)
Comment (mtrmac) [reply docs/dev/alternative_release_image_sources.md]: Actually, if `repositoryDigestMirrors` is supposed to be a consistent order, the `mirrors` list would also be the same, and include the primary one:
```
[[registry]]location = A; mirrors = A, B, C, D
[[registry]]location = B; mirrors = A, B, C, D
[[registry]]location = C; mirrors = A, B, C, D
[[registry]]location = D; mirrors = A, B, C, D
```
(and `registries.conf` would be modified not to contact the primary one again if it is already listed in mirrors.)

---

WRT this PR:
- `ReleaseImageSources` should _document_ whether `repository` or `sources.repository` is supposed to be accessed first. Assuming `repository` is the internet-accessible pull spec and `sources.repository` is the list of. mirrors, presumably the mirrors should be tried first.
- The example of generated `repositoryDigestMirrors` (and the code generating it) should generate an array of repositories in the order in which they should be attempted; Again, assuming `repository` is the internet-accessible pull spec and `sources.repository` is the list of. mirrors, and that the mirrors should be tried first, the example above should have `local.registry.com` entries before `q.io` entries in each `sources` array.

File: docs/dev/alternative_release_image_sources.md [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+[[registry.mirror]]
+location = "local.registry.com/ocp/release"
+
+[[registry.mirror]]
+location = "backup.local.registry.com/ocp/release"
+```
+
+### oc adm release mirror
+
+The `release mirror` mirrors the a release-image and all other images referenced in the release-image to another repository and then provides user's details of setting up the `install-config.yaml` for the user.
```
Comment (umohnani8) [inline docs/dev/alternative_release_image_sources.md]: remove "a" after "mirrors the"

**Comment by abhinavdahiya** [comment]
```console
[core@adahiya-0-mbz67-bootstrap ~]$ ls -lah /etc/containers/registries.conf
-rw-------. 1 root root 924 Jun 28 23:14 /etc/containers/registries.conf

[core@adahiya-0-mbz67-bootstrap ~]$ cat /etc/containers/registries.conf
[[registry]]
location = "local.registry.com/ocp/release"
insecure = false
mirror-by-digest-only = true

[[registry.mirror]]
location = "registry.svc.ci.openshift.org/origin/release"
insecure = false
[[registry.mirror]]
location = "registry.svc.ci.openshift.org/origin/4.1-2019-06-27-223758"
insecure = false
[[registry]]
location = "registry.svc.ci.openshift.org/origin/release"
insecure = false
mirror-by-digest-only = true

[[registry.mirror]]
location = "local.registry.com/ocp/release"
insecure = false
[[registry.mirror]]
location = "registry.svc.ci.openshift.org/origin/4.1-2019-06-27-223758"
insecure = false
[[registry]]
location = "registry.svc.ci.openshift.org/origin/4.1-2019-06-27-223758"
insecure = false
mirror-by-digest-only = true

[[registry.mirror]]
location = "local.registry.com/ocp/release"
insecure = false
[[registry.mirror]]
location = "registry.svc.ci.openshift.org/origin/release"
insecure = false
```
@umohnani8 
1. does the permissions on that file look all right?
2. does the content look alright?

```console
[core@adahiya-0-mbz67-bootstrap ~]$ journalctl -fu bootkube
-- Logs begin at Fri 2019-06-28 23:13:34 UTC. --
Jun 28 23:16:06 adahiya-0-mbz67-bootstrap systemd[1]: bootkube.service: Main process exited, code=exited, status=125/n/a
Jun 28 23:16:06 adahiya-0-mbz67-bootstrap systemd[1]: bootkube.service: Failed with result 'exit-code'.
Jun 28 23:16:12 adahiya-0-mbz67-bootstrap systemd[1]: bootkube.service: Service RestartSec=5s expired, scheduling restart.
Jun 28 23:16:12 adahiya-0-mbz67-bootstrap systemd[1]: bootkube.service: Scheduled restart job, restart counter is at 21.
Jun 28 23:16:12 adahiya-0-mbz67-bootstrap systemd[1]: Stopped Bootstrap a Kubernetes cluster.
Jun 28 23:16:12 adahiya-0-mbz67-bootstrap systemd[1]: Started Bootstrap a Kubernetes cluster.
Jun 28 23:16:12 adahiya-0-mbz67-bootstrap bootkube.sh[2799]: Pulling release image...
Jun 28 23:16:12 adahiya-0-mbz67-bootstrap bootkube.sh[2799]: error pulling image "registry.svc.ci.openshift.org/origin/release:4.2": unable to pull registry.svc.ci.openshift.org/origin/release:4.2: unable to pull image: Error initializing source docker://registry.svc.ci.openshift.org/origin/release:4.2: error loading registries: invalid URL: cannot be empty

[core@adahiya-0-mbz67-bootstrap ~]$ crio --version
crio version 1.13.9-1.rhaos4.1.gitd70609a.el8
```
@umohnani8 
Seeing ^^ this error on the bootstrap node, do we need a specific version of the CRI-O for v2 registries?

**Comment by umohnani8** [comment]
@abhinavdahiya v2 registry support was added in cri-o 1.14 https://github.com/cri-o/cri-o/pull/2494. We released cri-o 1.14 for OpenShift last week, I believe it should be in 4.2 soon, if not there already. @mrunalp do you know the timeline for cri-o 1.14 showing up in 4.2 clusters?

**Comment by mrunalp** [comment]
It is blocked on a RCM update of a dependency. Hopefully should be done tomorrow.

File: pkg/types/validation/installconfig.go [hunk 7]
```diff
@@ -268,3 +271,19 @@ func validateProxy(p *types.Proxy, fldPath *field.Path) field.ErrorList {
 
 	return allErrs
 }
+
+func validateReleaseImageSources(sources []types.ReleaseImageSource, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	for _, source := range sources {
+		ref, err := dockerref.ParseNamed(source.Repository)
```
Comment (mtrmac) [inline pkg/types/validation/installconfig.go]: This would be better with `ParseNormalizedNamed`.  `ParseNamed` is `ParseNormalizedNamed`+ a check that the input is in one specified form (i.e. you must say `docker.io/openshift/foo` instead of `openshift/foo`).

Sure, assuming the content is not going to be hosted on `docker.io`, this does not matter all that much for _this_ code.

It’s rare that you _need_ to forbid the alternative forms, but copy&pasting code across projects could cause the `ParseNamed` calls to spread to other user-facing code that _should_ accept other forms and use `ParseNormalizednamed`.

OTOH the tests do explicitly test for rejecting the implied-`docker.io` forms, so if this was intentional, fine.

---

BTW if this is going to be long-term supporting Windows/containerd / whole-registry mirrors, note that `Parse{Normalized,}Named` does not accept registry hostnames (without repository specification). Everything here says “repository”, so it’s OK right now.

File: data/data/bootstrap/files/etc/containers/registries.conf.template [hunk 8]
```diff
@@ -0,0 +1,13 @@
+{{ range $r := .Registries -}}
```
Comment (mtrmac) [inline data/data/bootstrap/files/etc/containers/registries.conf.template]: Couldn’t this just encode a `sysregistriesv2.V2RegistriesConf` object directly instead of hard-coding the current set of fields of `$r`?

File: pkg/types/validation/installconfig.go (hunk 7 above)
Comment (abhinavdahiya) [reply pkg/types/validation/installconfig.go]: So I think we do want to make sure people use complete names. no implicit names are allowed.

> BTW if this is going to be long-term supporting Windows/containerd / whole-registry mirrors, note that Parse{Normalized,}Named does not accept registry hostnames (without repository specification)

can you expand a little with examples of how this will break for windows...?

Comment (mtrmac) [reply pkg/types/validation/installconfig.go]: `reference.ParseNamed("a.example.com")` interprets the input as `docker.io/library/a.example.com` and rejects it with `"repository name must be canonical"`.

File: data/data/bootstrap/files/etc/containers/registries.conf.template (hunk 8 above)
Comment (mtrmac) [reply data/data/bootstrap/files/etc/containers/registries.conf.template]: … sorry, probably not trivially in the template language, without adding helpers at least.

File: pkg/types/validation/installconfig.go (hunk 7 above)
Comment (abhinavdahiya) [reply pkg/types/validation/installconfig.go]: > `reference.ParseNamed("a.example.com")` interprets the input as `docker.io/library/a.example.com` and rejects it with `"repository name must be canonical"`.

I think it is rejecting it correctly right?

Comment (mtrmac) [reply pkg/types/validation/installconfig.go]: Yes, “correctly” in the sense that the code wanted a repository in canonical form, the input was not a repository in canonical form.

Just pointing out that eventually we might need to support hostnames as well — but the validation can certainly be relaxed (or a separate, differently-validated field, can be added) later when/if that arises.

Comment (mtrmac) [reply pkg/types/validation/installconfig.go]: … FWIW the context is that CRI-O does not currently run on Windows, and containerd’s mirroring configuration is only at the registry (= host:port) resolution.  (Of course, this particular bit of configuration would be one of the smaller things that need changing, the whole mirroring process would need adapting for the idea that the mirror must mirror all of (the relevant parts of) the registry, not just a single repository.)

**Comment by umohnani8** [comment]
@abhinavdahiya cri-o 1.14 is in 4.2 now. Tests should pass now after rebase

**Comment by abhinavdahiya** [comment]
/hold

for allow review from multiple people...

File: docs/dev/alternative_release_image_sources.md (hunk 4 above)
Comment (smarterclayton) [reply docs/dev/alternative_release_image_sources.md]: I think in general mirrors should be preferred (we get an advantage in both cases) when specified.

File: pkg/asset/ignition/bootstrap/bootstrap.go [hunk 9]
```diff
@@ -165,31 +172,35 @@ func (a *Bootstrap) Files() []*asset.File {
...
-		if err != nil {
-			return nil, err
+	registries := []sysregistriesv2.Registry{}
+	if len(releaseImageSources) > 1 {
+		for i := 0; i < len(releaseImageSources); i++ {
+			registry := sysregistriesv2.Registry{}
+			registry.Endpoint.Location = releaseImageSources[i].Repository
+			registry.MirrorByDigestOnly = true
+			for j := 0; j < len(releaseImageSources); j++ {
+				if j == i {
```
Comment (mtrmac) [inline pkg/asset/ignition/bootstrap/bootstrap.go]: Non-blocking:

The M-C-O code does not have this condition; that way, the order of attempted mirrors is always the same regardless of which one was used in the pull spec. (+, yes, the `releaseImageSources[I].Repository` one is tried at the end, redundantly, but we will probably. change that in CRI-O eventually).

If you change this, please update `docs/dev/alternative_release_image_sources.md` as well.

File: docs/dev/alternative_release_image_sources.md [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+
+```go
+type InstallConfig struct {
+    ///
+
+    // ReleaseImageSources is the list of sources/repositories for pulling the release-image
+    // content.
+    // Any of the source/repository can be used to pull the release-image content.
+    // The order is the list has no meaning in terms of preference.
+    ReleaseImageSources []ReleaseImageSource `json:"releaseImageSources"`
```
Comment (smarterclayton) [inline docs/dev/alternative_release_image_sources.md]: So a key question here is whether we want people to provide initial content mirrors here.  I think a reasonable person could say this as cluster level setting (this set of content from A, this set of content for B).  I'm concerned that this is too specific (looks different from ICSP in a significant way).  

As an alternative, if we called this InstallContentSources and the structure matched RepositoryDigestMirrors, what is the impact?  This would be the initial set of ICSP objects - it would expose a subset of the fields there (anyone who wants more can always add their own ICSP, but that's a 10% case at best), and you should use this field to seed your initial mirrors for core cluster content.

I'm concerned that we may have a case where core content needs two sets of mirrors at some point (release digests A-J come from repository 1, digests K-Z come from repository 2), which would necessitate the array of arrays struct like we have for ICSP.  Also, today we have the samples operator (which in theory should be in OLM, but isn't, and won't be for a while) which won't come up properly unless you also mirror it.

Are we 100% positive we only need one mirror list, vs potentially needing 2 at any point.  That's the argument for array or array of arrays.

Comment (smarterclayton) [reply docs/dev/alternative_release_image_sources.md]: In general, looking more like ICSP (but with a better name and all extraneous stuff scraped out) lets us explain it better.

I.e.

```
InstallContentSources []InstallContentSource

type InstallContentSource struct {
  // only fields from ICSP.RepositoryDigestMirrors that make sense
  Sources []string
}
```

Comment (mtrmac) [reply docs/dev/alternative_release_image_sources.md]: > As an alternative, if we called this InstallContentSources and the structure matched RepositoryDigestMirrors, what is the impact?

The code would have to handle input where the two `InstallContentSources` arrays had common elements. (This is not a blocker - the code can either explicitly reject such input and tell the user to combine the lists, or use the code that already exists in machine-config-operator to combine them; just something to pay attention to.)

Comment (abhinavdahiya) [reply docs/dev/alternative_release_image_sources.md]: >  (release digests A-J come from repository 1, digests K-Z come from repository 2), which would necessitate the array of arrays struct like we have for ICSP.

with current ICSP api definition `[a,b,c] [c,d,e] => [a,b,c,d]` so how does multiple lists help with
`[release, release-image-refA-J, mirror] [release, release-image-refK-Z, mirror]` in the installer as that already means `[release, release-image-refA-J, release-image-refK-Z, mirror]`

@smarterclayton I don't completely understand the use-case for core-content.

and for extra content sources, i'm more incline to push that to day-2.. no reason that needs to be day-1

File: pkg/types/installconfig.go [hunk 11]
```diff
@@ -77,6 +77,10 @@ type InstallConfig struct {
 	// If unset, the cluster will not be configured to use a proxy.
 	// +optional
 	Proxy *Proxy `json:"proxy,omitempty"`
+
+	// ImageContentSources is the lists of sources/repositories that can be used to pull the same content.
```
Comment (smarterclayton) [inline pkg/types/installconfig.go]: Before we hit GA we should clean up this description, I think it's fine for now.

**Comment by smarterclayton** [comment]
API is fine, we can clean up godoc later.  Do you need deeper review than that?

**Comment by abhinavdahiya** [comment]
> API is fine, we can clean up godoc later. Do you need deeper review than that?

/hold cancel

**Comment by umohnani8** [comment]
@abhinavdahiya needs rebase.

**Comment by patrickdillon** [comment]
looks like this needs a rebase (but I don't see the label).

File: docs/dev/alternative_release_image_sources.md [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+
+The `release mirror` mirrors the a release-image and all other images referenced in the release-image to another repository and then provides user's details of setting up the `install-config.yaml` for the user.
+
+The output would look like:
+
+```console
+Release Image q.io/ocp/release-x.y@sha256:abcd was successfully mirrored to local.registry.com/ocp/release-x.y@sha256:abcd
+
+Following section can be added to the install-config.yaml to create a cluster using new repository:
+imageContentSources:
```
Comment (mtnbikenc) [inline docs/dev/alternative_release_image_sources.md]: Should this be `imageContentSource:`?  I'm seeing both in this code and I need to align with my PR for `oc`. https://github.com/openshift/origin/pull/23381

Comment (abhinavdahiya) [reply docs/dev/alternative_release_image_sources.md]: i think it needs to be `imageContentSources` based on the json tag on the actual install-config
https://github.com/openshift/installer/blame/bea51930a0e49b5da2fb58f2184ba27e79ceeb6c/docs/dev/alternative_release_image_sources.md#L29

Comment (mtnbikenc) [reply docs/dev/alternative_release_image_sources.md]: Thanks!  I'll update the mirror command.

File: docs/dev/alternative_release_image_sources.md (hunk 2 above)
Comment (mtnbikenc) [reply docs/dev/alternative_release_image_sources.md]: Implementing the changes to the mirror command here https://github.com/openshift/origin/pull/23381
